unixfs/io/sizeadj.go

```diff
@@ -0,0 +1,109 @@
+package io
+
+import (
+	"errors"
+	"io"
+)
+
+func newSizeAdjReadSeekCloser(base ReadSeekCloser, size uint64) *sizeAdjReadSeekCloser {
+	r := &sizeAdjReadSeekCloser{base: base, size: int64(size)}
+	if r.size < 0 {
+		panic("size limited 2^63 − 1")
+	}
+	return r
+}
+
+type sizeAdjReadSeekCloser struct {
+	base   ReadSeekCloser
+	size   int64
+	offset int64
+}
+
+// Read implements the Read method as defined by io.Reader
+func (r *sizeAdjReadSeekCloser) Read(p []byte) (int, error) {
+	if r.offset >= r.size { // EOF
+		return 0, io.EOF
+	}
+	if int64(len(p)) > r.size-r.offset { // truncate
+		newsize := r.size - r.offset
+		p = p[:newsize]
+	}
+	n, err := r.base.Read(p)
+	if err == nil {
+		_, err = r.base.Read(nil)
+	}
+	if err != io.EOF { // only pad when we get an EOF
+		r.offset += int64(n)
+		return n, err
+	}
+	for ; n < len(p) && r.offset+int64(n) < r.size; n++ { // pad
+		p[n] = 0
+	}
+	r.offset += int64(n)
+	return n, io.EOF
+}
+
+// Seek implements the Seek method as defined by io.Seeker
+func (r *sizeAdjReadSeekCloser) Seek(offset int64, whence int) (int64, error) {
+	switch whence {
+	case io.SeekStart:
+		r.offset = offset
+	case io.SeekCurrent:
+		r.offset += offset
+	case io.SeekEnd:
+		r.offset = r.size + offset
+	}
+	if r.offset < 0 {
+		return -1, errors.New("Seek will result in negative position")
```

Comment:
Shouldn't this rollback the offset (technically)?

Reply:
Yes, will fix.